lib/puppet/agent.rb

```diff
@@ -13,6 +15,10 @@ class Puppet::Agent
   require 'puppet/util/splayer'
   include Puppet::Util::Splayer
 
+  # Special exception class used to signal an agent run has timed out.
+  class RunTimeoutError < Exception
```

Comment:
We should probably either raise `Timeout::Error` (we raise it other [timeout sections](https://github.com/puppetlabs/puppet/blob/1baa0b189ee38d965dce239651db8b62688d7f9a/lib/puppet/provider/exec.rb#L51-L69)) or if we want a custom Error that other code sections will reference then I think we should put this in [error.rb](https://github.com/puppetlabs/puppet/blob/1baa0b189ee38d965dce239651db8b62688d7f9a/lib/puppet/error.rb) and follow the conventions there.

I don't think it would be bad to go with the latter for our timeout sections in general (throwing a custom Puppet error) but  for the scope of this PR I would probably just go with the former (use the Ruby built in that is raised elsewhere).

Reply:
I went with a unique error class to avoid other uses of Timeout trapping and handling a generic `Timeout::Error`. Similarly, I didn't put the class in `error.rb` as it isn't a "public API" kind of thing that other bits of code should be raising or trapping.

Reply:
Oh. You don't want to have any internal code be capable of catching this exception? In that case then, I believe you simply just shouldn't pass an exception class into `Timeout.timeout`.

Reply:
Yeah, looked at that too. However, I'm not sure that behavior is available in Ruby 1.9.3 and it also uses Ruby's throw+catch in strange convoluted ways that I still don't understand after an hour of looking at the implementation.

On the other hand, `Thread.raise exception` is something that I feel like I have an understanding of.

Reply:
The behavior is available in 1.9.3, as well as 1.8.7 (though pre-2.4.0 there isn't a third optional argument for a message, and 1.8.7 the method returns true if there was no timeout, not the value of the block being timed).

The crux of it is that if you pass a class, the block is evaluated in a begin/rescue, and if not, it's evaluated in a catch block with an instance of an internal error class passed in.

Reply:
One thing that has been rolling around in my head is that these failures are pretty silent from the perspective of PuppetDB or the PE Console. Using a custom exception may allow us to trap this event and ensure a report was fired off while also keeping the Timeout block coupled to the block that takes out the catalog lock.

Reply:
Thinking about this more I don't think we should allow code to rescue from this exception.

My thinking is that this timeout is really a catchall, something-weird-went-wrong-that-we-don't-understand-and-so-we-need-to-abort-everything, kinda event. I think at that point we shouldn't try any report handling, network calls, etc.

For better error handling/reporting I think we should dig into any reports of timeouts that are from our code and properly timeout those lower level calls.

Reply:
Fair enough. I'll re-factor it to use a plain Timeout without a custom exception class.